Routemap nb

patrasar opened the pull request:
This PR introduces the following changes:

Modifications to the existing library route-map yang definitions.
Introduction of yang definitions for bgpd, ospfd, ospf6d, and zebra route-map match and set clauses.
North-bound callbacks implementation for all the route-map yang elements.

Signed-off-by: NaveenThanikachalam nthanikachal@vmware.com
Signed-off-by: Sarita Patra saritap@vmware.com

louberger commented:
seeing yang related issues in bgp startup...

> 2020/11/10 08:49:53 BGP: [EC 100663318] YANG model "frr-bgp@*" "*@*"not embedded, trying external file
> 2020/11/10 08:49:53 BGP: [EC 100663318] YANG model "frr-bgp@*" "*@*"not embedded, trying external file
> 2020/11/10 08:49:53 BGP: [EC 100663318] YANG model "frr-bgp-types@*" "*@*"not embedded, trying external file
> 2020/11/10 08:49:53 BGP: [EC 100663318] YANG model "frr-bgp-types@*" "*@*"not embedded, trying external file
> 2020/11/10 08:49:53 BGP: [EC 100663318] YANG model "frr-bgp@2019-12-03" "frr-bgp-common-structure@*"not embedded, trying external file
> 2020/11/10 08:49:53 BGP: [EC 100663318] YANG model "frr-bgp@2019-12-03" "frr-bgp-common-structure@*"not embedded, trying external file
> 2020/11/10 08:49:53 BGP: [EC 100663318] YANG model "frr-bgp@2019-12-03" "frr-bgp-common@*"not embedded, trying external file
> 2020/11/10 08:49:53 BGP: [EC 100663318] YANG model "frr-bgp@2019-12-03" "frr-bgp-common@*"not embedded, trying external file
> 2020/11/10 08:49:53 BGP: [EC 100663318] YANG model "frr-bgp@2019-12-03" "frr-bgp-common-multiprotocol@*"not embedded, trying external file
> 2020/11/10 08:49:53 BGP: [EC 100663318] YANG model "frr-bgp@2019-12-03" "frr-bgp-common-multiprotocol@*"not embedded, trying external file
> 2020/11/10 08:49:53 BGP: [EC 100663318] YANG model "frr-bgp@2019-12-03" "frr-bgp-neighbor@*"not embedded, trying external file
> 2020/11/10 08:49:53 BGP: [EC 100663318] YANG model "frr-bgp@2019-12-03" "frr-bgp-neighbor@*"not embedded, trying external file
> 2020/11/10 08:49:53 BGP: [EC 100663318] YANG model "frr-bgp@2019-12-03" "frr-bgp-peer-group@*"not embedded, trying external file
> 2020/11/10 08:49:53 BGP: [EC 100663318] YANG model "frr-bgp@2019-12-03" "frr-bgp-peer-group@*"not embedded, trying external file
> 2020/11/10 08:49:53 BGP: [EC 100663318] YANG model "frr-bgp@2019-12-03" "frr-bgp-bmp@*"not embedded, trying external file
> 2020/11/10 08:49:53 BGP: [EC 100663318] YANG model "frr-bgp@2019-12-03" "frr-bgp-bmp@*"not embedded, trying external file
> 2020/11/10 08:49:53 BGP: libyang: Failed to resolve uses "global-group-use-multiple-paths". (/frr-bgp:{grouping}[mp-afi-unicast-common]/global-group-use-multiple-paths)
> 2020/11/10 08:49:53 BGP: libyang: Failed to resolve uses "global-group-use-multiple-paths". (/frr-bgp:{grouping}[mp-afi-unicast-common]/global-group-use-multiple-paths)
> 2020/11/10 08:49:53 BGP: libyang: Failed to resolve uses "global-redistribute". (/frr-bgp:{grouping}[mp-afi-unicast-common]/global-redistribute)
> 2020/11/10 08:49:53 BGP: libyang: Failed to resolve uses "global-redistribute". (/frr-bgp:{grouping}[mp-afi-unicast-common]/global-redistribute)
> 2020/11/10 08:49:53 BGP: libyang: Failed to resolve uses "admin-distance". (/frr-bgp:{grouping}[mp-afi-unicast-common]/admin-distance)
> 2020/11/10 08:49:53 BGP: libyang: Failed to resolve uses "admin-distance". (/frr-bgp:{grouping}[mp-afi-unicast-common]/admin-distance)
> 2020/11/10 08:49:53 BGP: libyang: Failed to resolve uses "med-config". (/frr-routing:routing/control-plane-protocols/control-plane-protocol/frr-bgp:bgp/frr-bgp:global/frr-bgp:med-config)
> ...

rzalamena commented on bgpd/bgp_routemap.c:
```diff
@@ -4122,778 +4151,1079 @@ DEFUN (no_match_evpn_route_type,
+		    || (acln >= 1300 && acln <= 1999)) {
+			snprintf(
+				xpath_value, sizeof(xpath_value),
+				"%s/rmap-match-condition/frr-bgp-route-map:access-list-num",
```
This needs an update: access list is now only a string.

- https://github.com/FRRouting/frr/blob/master/yang/frr-route-map.yang#L264
- https://github.com/FRRouting/frr/blob/master/yang/frr-filter.yang#L109

patrasar replied:
DEFPY(match_ip_route_source, match_ip_route_source_cmd,
      "match ip route-source <(1-199)$acll|(1300-2699)$aclh|WORD$name>",
      MATCH_STR IP_STR
      "Match advertising source address of route\n"
      "IP access-list number\n"
      "IP access-list number (expanded range)\n"
      "IP standard access-list name\n")

We are maintaining 3 different types.
access-list-num for 1-99
access-list-num-extended 1300-269
list-name for word

idryzhov replied:
This one is still not addressed. As Rafael pointed out, the access-list is referenced by its name.
It doesn't matter that this name may actually be a set of characters or a number. It's still a string.
Instead of using three different leafs - `access-list-num`, `access-list-num-extended` and `access-list-name` just use single `access-list-name` for all three cases.

rzalamena commented on lib/routemap.h:
```diff
@@ -241,6 +241,146 @@ struct route_map {
 DECLARE_QOBJ_TYPE(route_map)
 
+/* Route-map match conditions */
+#define IS_MATCH_INTERFACE(C)                                                  \
```
This won't work as you expect:
- `strncmp` does not have the same functionality as `strcmp`
- `enum` will be more performant and will avoid all these macros

Please change this.

patrasar replied:
Library is not ready for identity-ref, so we are using strcmp. Once the code is pushed for Identity-ref we will replace these with enum

idryzhov replied:
@patrasar Rafael is right about `strncmp`.

For example, if you pass `"frr-bgp-route-map:ip-route-source"`, both `IS_MATCH_ROUTE_SRC` and `IS_MATCH_ROUTE_SRC_PL` will return `true`.

Let's replace it with `strmatch` like it's done in other places where we use identityrefs.

patrasar replied:
Done

rzalamena commented on lib/routemap_cli.c:
```diff
@@ -155,6 +155,28 @@ void route_map_instance_show(struct vty *vty, struct lyd_node *dnode,
 		SKIP_RULE("ip next-hop prefix-len");
 		SKIP_RULE("source-protocol");
 		SKIP_RULE("source-instance");
+		/* OSPF specific match conditions. */
```
If we have all route maps migrated, then this function is no longer necessary.

patrasar replied:
Addressed

rzalamena commented on lib/routemap_cli.c:
```diff
@@ -163,10 +185,38 @@ void route_map_instance_show(struct vty *vty, struct lyd_node *dnode,
 	/* Print route map `set` for old CLI users. */
 	for (rmr = rmi->set_list.head; rmr; rmr = rmr->next) {
 		/* Skip all sets implemented by northbound. */
+		SKIP_RULE("ip next-hop");
```
If we have all route maps migrated, then this function is no longer necessary.

patrasar replied:
Addressed

rzalamena reviewed (changes requested):
Seems like this is a clone of #6369 , should we close the old PR?

Also I think this PR got too big... maybe we should split it to help testing/merging faster? It will take a while to test all daemons each iteration of fixes here, maybe a topotest could help reduce the burden.

patrasar commented:
Working on the comments.

rwestphal reviewed (commented):
Since this PR is converting all remaining route-map commands to YANG, it should be possible to remove the following code from the base "route-map" command:
```diff
--- a/lib/routemap_cli.c
+++ b/lib/routemap_cli.c
@@ -66,17 +66,9 @@ DEFPY_YANG_NOSH(
        nb_cli_enqueue_change(vty, xpath_action, NB_OP_MODIFY, action);
 
        rv = nb_cli_apply_changes(vty, NULL);
-       if (rv == CMD_SUCCESS) {
+       if (rv == CMD_SUCCESS)
                VTY_PUSH_XPATH(RMAP_NODE, xpath_index);
 
-               /* Add support for non-migrated route map users. */
-               nb_cli_pending_commit_check(vty);
-               rm = route_map_get(name);
-               action_type = (action[0] == 'p') ? RMAP_PERMIT : RMAP_DENY;
-               rmi = route_map_index_get(rm, action_type, sequence);
-               VTY_PUSH_CONTEXT(RMAP_NODE, rmi);
-       }
-
        return rv;
 }
```
Removing that code will allow the YANG back-off timer hack (#6727) to work better since the pending configuration changes won't need to be committed (`nb_cli_pending_commit_check()`) every time a "route-map" command is parsed. That should have a huge impact in configuration loading times when a large number of route-maps is present (which isn't unusual).

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/d08bf0bf7072e0fea4c3ac3303ce1f80/raw/641bc713a1e84588bbfbe9bf083e774bcf24a634/cr_7419_1608265357.diff | git apply
```

</p>

<p>

```diff
diff --git a/bgpd/bgp_main.c b/bgpd/bgp_main.c
index 1bd67f35d..02123ea18 100644
--- a/bgpd/bgp_main.c
+++ b/bgpd/bgp_main.c
@@ -377,12 +377,8 @@ static void bgp_vrf_terminate(void)
 }
 
 static const struct frr_yang_module_info *const bgpd_yang_modules[] = {
-	&frr_bgp_info,
-	&frr_filter_info,
-	&frr_interface_info,
-	&frr_route_map_info,
-	&frr_routing_info,
-	&frr_vrf_info,
+	&frr_bgp_info,		 &frr_filter_info,  &frr_interface_info,
+	&frr_route_map_info,     &frr_routing_info, &frr_vrf_info,
 	&frr_bgp_route_map_info,
 };
 
diff --git a/bgpd/bgp_routemap.c b/bgpd/bgp_routemap.c
index 0ce4570ea..6576eef85 100644
--- a/bgpd/bgp_routemap.c
+++ b/bgpd/bgp_routemap.c
@@ -5672,12 +5672,10 @@ DEFPY (match_ipv4_next_hop,
 	char xpath_value[XPATH_MAXLEN];
 
 	nb_cli_enqueue_change(vty, xpath, NB_OP_CREATE, NULL);
-	snprintf(
-		xpath_value, sizeof(xpath_value),
-		"%s/rmap-match-condition/frr-bgp-route-map:ipv4-address",
-		xpath);
-	nb_cli_enqueue_change(vty, xpath_value, NB_OP_MODIFY,
-			      argv[4]->arg);
+	snprintf(xpath_value, sizeof(xpath_value),
+		 "%s/rmap-match-condition/frr-bgp-route-map:ipv4-address",
+		 xpath);
+	nb_cli_enqueue_change(vty, xpath_value, NB_OP_MODIFY, argv[4]->arg);
 
 	return nb_cli_apply_changes(vty, NULL);
 }
diff --git a/lib/routemap_cli.c b/lib/routemap_cli.c
index c50fadbe8..96022f21c 100644
--- a/lib/routemap_cli.c
+++ b/lib/routemap_cli.c
@@ -788,8 +788,8 @@ void route_map_condition_show(struct vty *vty, struct lyd_node *dnode,
 
 			if (true
 			    == yang_dnode_get_bool(
-					dnode,
-					"./rmap-match-condition/frr-bgp-route-map:comm-list-standard/comm-list-num-exact-match"))
+				    dnode,
+				    "./rmap-match-condition/frr-bgp-route-map:comm-list-standard/comm-list-num-exact-match"))
 				vty_out(vty,
 					" match community %s exact-match\n",
 					acl);
@@ -804,8 +804,8 @@ void route_map_condition_show(struct vty *vty, struct lyd_node *dnode,
 
 			if (true
 			    == yang_dnode_get_bool(
-					dnode,
-					"./rmap-match-condition/frr-bgp-route-map:comm-list-extended/comm-list-num-extended-exact-match"))
+				    dnode,
+				    "./rmap-match-condition/frr-bgp-route-map:comm-list-extended/comm-list-num-extended-exact-match"))
 				vty_out(vty,
 					" match community %s exact-match\n",
 					acl);
@@ -821,8 +821,8 @@ void route_map_condition_show(struct vty *vty, struct lyd_node *dnode,
 
 			if (true
 			    == yang_dnode_get_bool(
-					dnode,
-					"./rmap-match-condition/frr-bgp-route-map:comm-list/comm-list-name-exact-match"))
+				    dnode,
+				    "./rmap-match-condition/frr-bgp-route-map:comm-list/comm-list-name-exact-match"))
 				vty_out(vty,
 					" match community %s exact-match\n",
 					acl);
@@ -841,8 +841,8 @@ void route_map_condition_show(struct vty *vty, struct lyd_node *dnode,
 
 			if (true
 			    == yang_dnode_get_bool(
-					dnode,
-					"./rmap-match-condition/frr-bgp-route-map:comm-list-standard/comm-list-num-exact-match"))
+				    dnode,
+				    "./rmap-match-condition/frr-bgp-route-map:comm-list-standard/comm-list-num-exact-match"))
 				vty_out(vty,
 					" match large-community %s exact-match\n",
 					acl);
@@ -858,8 +858,8 @@ void route_map_condition_show(struct vty *vty, struct lyd_node *dnode,
 
 			if (true
 			    == yang_dnode_get_bool(
-					dnode,
-					"./rmap-match-condition/frr-bgp-route-map:comm-list-extended/comm-list-num-extended-exact-match"))
+				    dnode,
+				    "./rmap-match-condition/frr-bgp-route-map:comm-list-extended/comm-list-num-extended-exact-match"))
 				vty_out(vty,
 					" match large-community %s exact-match\n",
 					acl);
@@ -876,8 +876,8 @@ void route_map_condition_show(struct vty *vty, struct lyd_node *dnode,
 
 			if (true
 			    == yang_dnode_get_bool(
-					dnode,
-					"./rmap-match-condition/frr-bgp-route-map:comm-list/comm-list-name-exact-match"))
+				    dnode,
+				    "./rmap-match-condition/frr-bgp-route-map:comm-list/comm-list-name-exact-match"))
 				vty_out(vty,
 					" match large-community %s exact-match\n",
 					acl);
@@ -1124,7 +1124,9 @@ void route_map_action_show(struct vty *vty, struct lyd_node *dnode,
 			vty_out(vty, " set metric +%s\n",
 				yang_dnode_get_string(
 					dnode, "./rmap-set-action/add-metric"));
-		} else if (yang_dnode_get(dnode, "./rmap-set-action/subtract-metric")) {
+		} else if (yang_dnode_get(
+				   dnode,
+				   "./rmap-set-action/subtract-metric")) {
 			vty_out(vty, " set metric -%s\n",
 				yang_dnode_get_string(
 					dnode,
@@ -1254,8 +1256,8 @@ void route_map_action_show(struct vty *vty, struct lyd_node *dnode,
 		else {
 			if (true
 			    == yang_dnode_get_bool(
-				       dnode,
-				       "./rmap-set-action/frr-bgp-route-map:large-community-none"))
+				    dnode,
+				    "./rmap-set-action/frr-bgp-route-map:large-community-none"))
 				vty_out(vty, " set large-community none\n");
 		}
 	} else if (IS_SET_COMMUNITY(action)) {
@@ -1269,8 +1271,8 @@ void route_map_action_show(struct vty *vty, struct lyd_node *dnode,
 		else {
 			if (true
 			    == yang_dnode_get_bool(
-				       dnode,
-				       "./rmap-set-action/frr-bgp-route-map:community-none"))
+				    dnode,
+				    "./rmap-set-action/frr-bgp-route-map:community-none"))
 				vty_out(vty, " set community none\n");
 		}
 	} else if (IS_SET_EXTCOMMUNITY_RT(action)) {
@@ -1323,14 +1325,14 @@ void route_map_action_show(struct vty *vty, struct lyd_node *dnode,
 	} else if (IS_SET_IPV6_PEER_ADDR(action)) {
 		if (true
 		    == yang_dnode_get_bool(
-			       dnode,
-			       "./rmap-set-action/frr-bgp-route-map:preference"))
+			    dnode,
+			    "./rmap-set-action/frr-bgp-route-map:preference"))
 			vty_out(vty, " set ipv6 next-hop peer-address\n");
 	} else if (IS_SET_IPV6_PREFER_GLOBAL(action)) {
 		if (true
 		    == yang_dnode_get_bool(
-			       dnode,
-			       "./rmap-set-action/frr-bgp-route-map:preference"))
+			    dnode,
+			    "./rmap-set-action/frr-bgp-route-map:preference"))
 			vty_out(vty, " set ipv6 next-hop prefer-global\n");
 	} else if (IS_SET_IPV4_VPN_NH(action)) {
 		vty_out(vty, " set ipv4 vpn next-hop %s\n",
diff --git a/ospf6d/ospf6_main.c b/ospf6d/ospf6_main.c
index 0ee31a0e0..bd59730f1 100644
--- a/ospf6d/ospf6_main.c
+++ b/ospf6d/ospf6_main.c
@@ -168,11 +168,8 @@ struct quagga_signal_t ospf6_signals[] = {
 };
 
 static const struct frr_yang_module_info *const ospf6d_yang_modules[] = {
-	&frr_filter_info,
-	&frr_interface_info,
-	&frr_route_map_info,
-	&frr_vrf_info,
-	&frr_ospf_route_map_info,
+	&frr_filter_info, &frr_interface_info,      &frr_route_map_info,
+	&frr_vrf_info,    &frr_ospf_route_map_info,
 };
 
 FRR_DAEMON_INFO(ospf6d, OSPF6, .vty_port = OSPF6_VTY_PORT,
diff --git a/ospfd/ospf_main.c b/ospfd/ospf_main.c
index ec0e651a5..cd1b0010b 100644
--- a/ospfd/ospf_main.c
+++ b/ospfd/ospf_main.c
@@ -128,11 +128,8 @@ struct quagga_signal_t ospf_signals[] = {
 };
 
 static const struct frr_yang_module_info *const ospfd_yang_modules[] = {
-	&frr_filter_info,
-	&frr_interface_info,
-	&frr_route_map_info,
-	&frr_vrf_info,
-	&frr_ospf_route_map_info,
+	&frr_filter_info, &frr_interface_info,      &frr_route_map_info,
+	&frr_vrf_info,    &frr_ospf_route_map_info,
 };
 
 FRR_DAEMON_INFO(ospfd, OSPF, .vty_port = OSPF_VTY_PORT,
diff --git a/zebra/main.c b/zebra/main.c
index ef983969d..46631d512 100644
--- a/zebra/main.c
+++ b/zebra/main.c
@@ -255,12 +255,8 @@ struct quagga_signal_t zebra_signals[] = {
 };
 
 static const struct frr_yang_module_info *const zebra_yang_modules[] = {
-	&frr_filter_info,
-	&frr_interface_info,
-	&frr_route_map_info,
-	&frr_zebra_info,
-	&frr_vrf_info,
-	&frr_routing_info,
+	&frr_filter_info,	  &frr_interface_info, &frr_route_map_info,
+	&frr_zebra_info,	   &frr_vrf_info,       &frr_routing_info,
 	&frr_zebra_route_map_info,
 };
 

```

</p>
</details>

            
If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

patrasar commented:
> Since this PR is converting all remaining route-map commands to YANG, it should be possible to remove the following code from the base "route-map" command:
> 
> ```diff
> --- a/lib/routemap_cli.c
> +++ b/lib/routemap_cli.c
> @@ -66,17 +66,9 @@ DEFPY_YANG_NOSH(
>         nb_cli_enqueue_change(vty, xpath_action, NB_OP_MODIFY, action);
>  
>         rv = nb_cli_apply_changes(vty, NULL);
> -       if (rv == CMD_SUCCESS) {
> +       if (rv == CMD_SUCCESS)
>                 VTY_PUSH_XPATH(RMAP_NODE, xpath_index);
>  
> -               /* Add support for non-migrated route map users. */
> -               nb_cli_pending_commit_check(vty);
> -               rm = route_map_get(name);
> -               action_type = (action[0] == 'p') ? RMAP_PERMIT : RMAP_DENY;
> -               rmi = route_map_index_get(rm, action_type, sequence);
> -               VTY_PUSH_CONTEXT(RMAP_NODE, rmi);
> -       }
> -
>         return rv;
>  }
> ```
> 
> Removing that code will allow the YANG back-off timer hack (#6727) to work better since the pending configuration changes won't need to be committed (`nb_cli_pending_commit_check()`) every time a "route-map" command is parsed. That should have a huge impact in configuration loading times when a large number of route-maps is present (which isn't unusual).

Addressed

patrasar commented:
ci:rerun

patrasar commented:
I have addressed all the comments.
Rebasing code causes multiple test cases failure, i am investigating the reason.

patrasar commented:
ci:rerun

idryzhov commented:
1. Rafael's comment is not addressed: https://github.com/FRRouting/frr/pull/7419#discussion_r521316769
2. You're using "\n" in some "when" statements. This is not prohibited, but in all other yang models I saw, multi-line strings are concatenated with "+" instead.
3. Why isn't "[no] match script WORD" command converted?
4. All converted DEFUN/DEFPY should be changed to DEFUN_YANG/DEFPY_YANG.
5. interface match-condition should use "type frr-interface:interface-ref" instead of a simple string.

Other than that, I think it is not possible to really review such an amount of code.
As we pushed back the release and waiting for libyang2, I suppose it should be merged ASAP to have time to test it.

patrasar commented:
> 1. Rafael's comment is not addressed: [#7419 (comment)](https://github.com/FRRouting/frr/pull/7419#discussion_r521316769)
> 2. You're using "\n" in some "when" statements. This is not prohibited, but in all other yang models I saw, multi-line strings are concatenated with "+" instead.
> 3. Why isn't "[no] match script WORD" command converted?
> 4. All converted DEFUN/DEFPY should be changed to DEFUN_YANG/DEFPY_YANG.
> 5. interface match-condition should use "type frr-interface:interface-ref" instead of a simple string.
> 
> Other than that, I think it is not possible to really review such an amount of code.
> As we pushed back the release and waiting for libyang2, I suppose this PR should be merged ASAP to have time to test it.

Hi Igor,
Thanks for the comments.
I have addressed the comments. Please let me know if something i missed.

Regards,
Sarita

idryzhov commented on yang/frr-bgp-filter.yang:
```diff
@@ -0,0 +1,345 @@
+            "Extcommunity-list instance name/identifier";
+        }
+
+        choice Extcommunity-string {
```
Should start with lowercase letter.

patrasar replied:
Done

idryzhov commented on yang/frr-bgp-route-map.yang:
```diff
@@ -0,0 +1,855 @@
+    }
+
+    case access-list-name {
+      when "derived-from-or-self(/frr-route-map:lib/frr-route-map:route-map/frr-route-map:entry/frr-route-map:match-condition/frr-route-map:condition, 'mac-address-list') orderived-from-or-self(/frr-route-map:lib/frr-route-map:route-map/frr-route-map:entry/frr-route-map:match-condition/frr-route-map:condition, 'as-path-list') orderived-from-or-self(/frr-route-map:lib/frr-route-map:route-map/frr-r ...
```
No space after `or` (`orderived-from-or-self`). I wonder why yanglint doesn't complain about that... 
Also, please, make it multiple strings. Same comment for all similar issues.

patrasar replied:
Done

idryzhov commented on yang/frr-bgp-route-map.yang:
```diff
@@ -0,0 +1,855 @@
+      }
+    }
+
+    case comm-list-num {
```
I think we should do the same thing here as we did with access-lists.
Just use single `comm-list-name` case instead of three different cases.
Even in `frr-bgp-filter.yang` community-lists are addressed by name, not number.

patrasar replied:
Done

idryzhov commented on yang/frr-bgp-route-map.yang:
```diff
@@ -0,0 +1,855 @@
+      }
+    }
+
+    case comm-list-num {
```
Same thing about single `comm-list-name` case.

patrasar replied:
Done

idryzhov commented on lib/routemap_cli.c:
```diff
@@ -132,47 +132,6 @@ void route_map_instance_show(struct vty *vty, struct lyd_node *dnode,
 		 */
 		return;
 	}
-
```
`rmi = nb_running_get_entry(dnode, NULL, false);` above should be also removed.

patrasar replied:
Done

patrasar replied:
Hi @idryzhov,
Thanks for your comments.
I have addressed these and updated the patch

patrasar commented on bgpd/bgp_routemap.c:
```diff
@@ -3458,10 +3465,10 @@ static int bgp_route_match_add(struct vty *vty, const char *command,
 /* Delete bgp route map rule. */
-static int bgp_route_match_delete(struct vty *vty, const char *command,
-				  const char *arg, route_map_event_t type)
+int bgp_route_match_delete(struct vty *vty, struct route_map_index *index,
```
Done

patrasar commented on bgpd/bgp_routemap.c:
```diff
@@ -3430,21 +3431,27 @@ static const struct route_map_rule_cmd route_set_originator_id_cmd = {
 /* Add bgp route map rule. */
-static int bgp_route_match_add(struct vty *vty, const char *command,
-			       const char *arg, route_map_event_t type)
+int bgp_route_match_add(struct vty *vty, struct route_map_index *index,
```
Done

patrasar commented on bgpd/bgp_routemap.c:
```diff
@@ -3963,813 +3987,1234 @@ static const char *parse_evpn_rt_type(const char *num_rt_type)
+	int acln = acll ? acll : aclh;
+
+	nb_cli_enqueue_change(vty, xpath, NB_OP_CREATE, NULL);
+	if (name) {
```
Done

patrasar commented on bgpd/bgp_routemap.c:
```diff
@@ -40,6 +40,7 @@
 #include "frrstr.h"
 #include "network.h"
+#include "lib/northbound_cli.h"
 
```
Done

patrasar commented on bgpd/bgp_rpki.c:
```diff
@@ -943,7 +943,7 @@ static int config_write(struct vty *vty)
 }
 
-DEFUN_NOSH (rpki,
+DEFUN_YANG_NOSH (rpki,
```
Done

patrasar commented on bgpd/bgpd.h:
```diff
@@ -2367,4 +2368,11 @@ DECLARE_HOOK(bgp_snmp_update_stats,
 	     (rn, pi, added))
 void peer_nsf_stop(struct peer *peer);
 
+extern const struct frr_yang_module_info frr_bgp_route_map_info;
```
Done

patrasar commented on ospf6d/ospf6_asbr.c:
```diff
@@ -1722,99 +1723,83 @@ ospf6_routemap_rule_set_tag(void *rule, const struct prefix *p, void *object)
-       "Type of metric\n"
-       "OSPF6 external type 1 metric\n"
-       "OSPF6 external type 2 metric\n")
+DEFUN(ospf6_routemap_set_metric_type, ospf6_routemap_set_metric_type_cmd,
```
Done

patrasar commented on bgpd/bgp_routemap.c:
```diff
@@ -5373,119 +5960,136 @@ DEFPY (no_match_ipv4_next_hop,
+	return nb_cli_apply_changes(vty, NULL);
 }
 
 #ifdef KEEP_OLD_VPN_COMMANDS
```
Done

patrasar commented on ospf6d/ospf6d.h:
```diff
@@ -102,4 +102,5 @@ extern struct route_node *route_prev(struct route_node *node);
 extern void ospf6_debug(void);
 extern void ospf6_init(struct thread_master *master);
 
+extern const struct frr_yang_module_info frr_ospf_route_map_info;
```
Done

patrasar commented on ospfd/ospfd.h:
```diff
@@ -727,4 +727,6 @@ extern int p_spaces_compare_func(const struct p_space *a,
 extern int q_spaces_compare_func(const struct q_space *a,
 				 const struct q_space *b);
+
+extern const struct frr_yang_module_info frr_ospf_route_map_info;
```
Done

patrasar commented on yang/frr-bgp-filter.yang:
```diff
@@ -0,0 +1,345 @@
+          description
+            "Large community string";
+          case standard {
+            when "./type = 'large-community-list-standard-id' or"
```
Done

idryzhov commented:
@patrasar rebase on top of latest master to fix NetDEF-CI failures

patrasar commented:
> rebase on top of latest master to fix NetDEF-CI failures

Sure.

idryzhov commented on lib/routemap_cli.c:
```diff
@@ -800,58 +968,288 @@ DEFPY_YANG(
 	return nb_cli_apply_changes(vty, NULL);
 }
 
+DEFUN (set_srte_color,
```
DEFUN_YANG

patrasar replied:
Done

idryzhov commented on lib/routemap_cli.c:
```diff
@@ -800,58 +968,288 @@ DEFPY_YANG(
+        return nb_cli_apply_changes(vty, NULL);
+}
+
+DEFUN (no_set_srte_color,
```
DEFUN_YANG

patrasar replied:
Done

idryzhov commented on lib/routemap_cli.c:
```diff
@@ -800,58 +968,288 @@ DEFPY_YANG(
-		if (yang_dnode_exists(dnode, "./frr-zebra:source-v4"))
+			yang_dnode_get_string(dnode, "./rmap-set-action/tag"));
+	} else if (IS_SET_SR_TE_COLOR(action)) {
+		vty_out(vty, " set sr-te color  %s\n",
```
Double space after `color`.

patrasar replied:
Done

idryzhov commented on ospf6d/ospf6_main.c:
```diff
@@ -47,6 +47,7 @@
 #include "ospf6_lsa.h"
 #include "ospf6_interface.h"
 #include "ospf6_zebra.h"
+#include "../ospfd/ospf_routemap_nb.h"
```
Why do we include from `ospf`, not from `ospf6` here?

patrasar replied:
Changed to #include "ospf6_routemap_nb.h"

idryzhov commented on ospf6d/ospf6_main.c:
```diff
@@ -172,6 +173,7 @@ static const struct frr_yang_module_info *const ospf6d_yang_modules[] = {
 	&frr_interface_info,
 	&frr_route_map_info,
 	&frr_vrf_info,
+	&frr_ospf_route_map_info,
```
Let's rename to `frr_ospf6_route_map_info` to not confuse with ospfd structure.
(Also in `ospf6d/ospf6_routemap_nb.c/h`)

patrasar replied:
Done

idryzhov commented on yang/frr-ospf-route-map.yang:
```diff
@@ -0,0 +1,65 @@
+    "FRR Users List:       <mailto:frog@lists.frrouting.org>
+     FRR Development List: <mailto:dev@lists.frrouting.org>";
+  description
+    "This module defines bgp route map settings";
```
1. `bgp` -> `ospf`
2. Should we have a separate yang module for ospf6? ospf6 implements `forwarding-address` while ospf does not.

patrasar replied:
Done

idryzhov commented on yang/frr-zebra-route-map.yang:
```diff
@@ -0,0 +1,125 @@
+    "FRR Users List:       <mailto:frog@lists.frrouting.org>
+     FRR Development List: <mailto:dev@lists.frrouting.org>";
+  description
+    "This module defines bgp route map settings";
```
`bgp` -> `zebra`

patrasar replied:
Done

idryzhov commented:
@patrasar Thanks for your work! I left a couple of comments for some minor issues. Once they are fixed I think this PR is ready to be merged

patrasar commented on bgpd/bgp_rpki.c:
```diff
@@ -953,7 +953,7 @@ DEFUN_NOSH (rpki,
 	return CMD_SUCCESS;
 }
 
-DEFUN (bgp_rpki_start,
```
Done

idryzhov replied:
This one is not addressed

patrasar replied:
HI Igor,
Somehow the changes were not saved, pushed the changes again.
Thank you.

patrasar commented:
@idryzhov Hi Igor, 
Thanks for the comments. I have addressed all of those.

idryzhov reviewed (approved):
There are some style issues reported by CI, I don't know if this is critical.
Otherwise LGTM.

patrasar commented:
> Since this PR is converting all remaining route-map commands to YANG, it should be possible to remove the following code from the base "route-map" command:
> 
> ```diff
> --- a/lib/routemap_cli.c
> +++ b/lib/routemap_cli.c
> @@ -66,17 +66,9 @@ DEFPY_YANG_NOSH(
>         nb_cli_enqueue_change(vty, xpath_action, NB_OP_MODIFY, action);
>  
>         rv = nb_cli_apply_changes(vty, NULL);
> -       if (rv == CMD_SUCCESS) {
> +       if (rv == CMD_SUCCESS)
>                 VTY_PUSH_XPATH(RMAP_NODE, xpath_index);
>  
> -               /* Add support for non-migrated route map users. */
> -               nb_cli_pending_commit_check(vty);
> -               rm = route_map_get(name);
> -               action_type = (action[0] == 'p') ? RMAP_PERMIT : RMAP_DENY;
> -               rmi = route_map_index_get(rm, action_type, sequence);
> -               VTY_PUSH_CONTEXT(RMAP_NODE, rmi);
> -       }
> -
>         return rv;
>  }
> ```
> 
> Removing that code will allow the YANG back-off timer hack (#6727) to work better since the pending configuration changes won't need to be committed (`nb_cli_pending_commit_check()`) every time a "route-map" command is parsed. That should have a huge impact in configuration loading times when a large number of route-maps is present (which isn't unusual).

Done

louberger commented:
from labn ci failure (I don't think it's a real core)
            
core_handler: memstats:  Route map compiled            :      1 *          7
core_handler: memstats:  Route map dependency          :      1 *         24
core_handler: memstats:  Route map dependency data     :      1 *         16
core_handler: memstats:  Skip List                     :     18 *         56
core_handler: memstats:  Skip Node                     :     37 * (variably sized)
core_handler: memstats:  Socket union                  :      6 *         28
core_handler: memstats:  Stream                        :     11 * (variably sized)
core_handler: memstats:  Stream FIFO                   :      6 *         64
core_handler: memstats:  Route table                   :    437 *         56
core_handler: memstats:  Route node                    :      1 *        112
core_handler: memstats:  Thread                        :     25 *        152
core_handler: memstats:  Thread master                 :     12 * (variably sized)
core_handler: memstats:  Thread Poll Info              :      6 *       8192
core_handler: memstats:  Thread stats                  :     32 *         72
core_handler: memstats:  Typed-hash bucket             :      8 * (variably sized)
core_handler: memstats:  Typed-heap array              :      1 * (variably sized)
core_handler: memstats:  Vector                        :  27794 *         16
core_handler: memstats:  Vector index                  :  27794 * (variably sized)
core_handler: memstats:  VRF                           :      1 *        216
core_handler: memstats:  VRF bit-map                   :      1 *          8
core_handler: memstats:  VTY                           :      2 * (variably sized)
core_handler: memstats:  VTY history                   :     20 * (variably sized)
core_handler: memstats:  Work queue                    :      9 *        152
core_handler: memstats:  Work queue name string        :      9 * (variably sized)
core_handler: memstats:  YANG module                   :      7 *         48
core_handler: memstats:  Zclient                       :      2 *       3480
core_handler: memstats:  Redistribution instance IDs   :      6 *          2
core_handler: memstats:  log thread-local buffer       :      2 *      10264
core_handler: showing active allocations in memory group logging subsystem
core_handler: memstats:  log file target               :      2 *         88
